src/util/svg-text-bubble.js

```diff
@@ -0,0 +1,176 @@
+              />
+              <path
+                d="${pathString}"
+                stroke="none"
+                fill="white" />
+            </g>`;
+    }
+
+    _thinkBubble (w, h, radius, pointsLeft) {
+        const e1rx = 2.25;
+        const e1ry = 2.25;
+        const e2rx = 1.5;
+        const e2ry = 1.5;
+        const e1x = 16 + 7 + e1rx;
+        const e1y = 5 + h + e1ry;
+        const e2x = 16 + e2rx;
+        const e2y = 8 + h + e2ry;
+        const insetR = 4;
+        const pInset1 = 12 + radius;
+        const pInset2 = pInset1 + (2 * insetR);
+
+        let pathString = `
+            M 0 ${radius}
+            A ${radius} ${radius} 0 0 1 ${radius} 0
+            L ${w - radius} 0
+            A ${radius} ${radius} 0 0 1 ${w} ${radius}
+            L ${w} ${h - radius}
+            A ${radius} ${radius} 0 0 1 ${w - radius} ${h}`;
+
+        if (pointsLeft) {
+            pathString += `
+                L ${pInset2} ${h}
+                A ${insetR} ${insetR} 0 0 1 ${pInset2 - insetR} ${h + insetR}
+                A ${insetR} ${insetR} 0 0 1 ${pInset1} ${h}`;
+        } else {
+            pathString += `
+                L ${w - pInset1} ${h}
+                A ${insetR} ${insetR} 0 0 1 ${w - pInset1 - insetR} ${h + insetR}
+                A ${insetR} ${insetR} 0 0 1 ${w - pInset2} ${h}`;
+        }
+
+        pathString += `
+            L ${radius} ${h}
+            A ${radius} ${radius} 0 0 1 0 ${h - radius}
+            Z`;
+
+        const ellipseSvg = (cx, cy, rx, ry) => `
+            <g>
+                <ellipse
+                    cx="${cx}" cy="${cy}"
+                    rx="${rx}" ry="${ry}"
+                    fill="rgba(0, 0, 0, 0.15)"
+                    stroke="rgba(0, 0, 0, 0.15)"
+                    stroke-width="4"
+                />
+                <ellipse
+                    cx="${cx}" cy="${cy}"
+                    rx="${rx}" ry="${ry}"
+                    fill="white"
+                    stroke="none"
+                />
+            </g>`;
+        let ellipses = [];
+        if (pointsLeft) {
+            ellipses = [
+                ellipseSvg(e1x, e1y, e1rx, e1ry),
+                ellipseSvg(e2x, e2y, e2rx, e2ry)
+            ];
+        } else {
+            ellipses = [
+                ellipseSvg(w - e1x, e1y, e1rx, e1ry),
+                ellipseSvg(w - e2x, e2y, e2rx, e2ry)
+            ];
+        }
+
+        return `
+             <g>
+                <path d="${pathString}" stroke="rgba(0, 0, 0, 0.15)" stroke-width="4" fill="rgba(0, 0, 0, 0.15)" />
+                <path d="${pathString}" stroke="none" fill="white" />
+                ${ellipses.join('\n')}
+            </g>`;
+    }
+
+
+    _getTextSize () {
+        const svgString = this._wrapSvgFragment(this._textFragment());
+        if (!this._textSizeCache[svgString]) {
+            this._textSizeCache[svgString] = this.svgRenderer.measure(svgString);
+        }
+        return this._textSizeCache[svgString];
```

Comment:
Considering that the measurement provider also caches its results, this feels almost redundant to me. I recognize that it's a bit different, though. Maybe as a future enhancement we can find a way to only cache in one place.

Reply:
Yeah, unfortunately they are slightly different. I went back and forth trying to figure out a way to combine them, which would also help because we are measuring nearly the same thing twice in the non-cached case. But yeah i think that would be a good future enhancement.